File: Extras/nengo/takeoff.py
#!/usr/bin/env python3
'''
Test simple altitude-hold PID controller

Copyright (C) 2019 Simon D. Levy

MIT License
'''

from time import sleep
import numpy as np
from multicopter import Multicopter
from sys import stdout

# Target 
ALTITUDE_TARGET = 10

# PID params
KP = 0.4125
KD = 4.5

def debug(msg):

    stdout.write(msg)
    stdout.flush()

if __name__ == '__main__':

    # initial conditions
    z = 0
    u = 0
    tprev = 0

    debug('Importing Nengo ... ')
    from nengo_pid_controller import NengoPidController
    debug('done\n')

    # Create PID controller
    npid = NengoPidController(KP, KD)

    # Create a multicopter simulation
    copter = Multicopter()

    # Start the simulation
    copter.start()

    # Loop until user hits the stop button
    while True:

        # Get vehicle state from sim
        telem = copter.getState()

        # Extract time from state
        t =  telem[0]

        # Negative time means user hit stop button
        if t < 0: break

        # Extract altitude from state.  Altitude is in NED coordinates, so we negate it to use as input
        # to PID controller.
        z = -telem[9]

        # Compute vertical climb rate as first difference of altitude over time
        if (not np.isnan(z)) and (abs(z) < 100) and (t > tprev):

            debug('z=%+3.3f  u=%3.3f\n' % (z, u))

            # Get correction from PID controller
            u = npid.getCorrection(ALTITUDE_TARGET, z)[0]

            # Constrain correction to [0,1] to represent motor value
            u = max(0, min(1, u))

        # Set motor values in sim
        copter.setMotors(u*np.ones(4))

        # Update for first difference
        tprev = t

        # Yield to Multicopter thread
        sleep(.001)

File: Extras/servers/java/Takeoff.java
/*
   Java Multicopter test 

   Copyright(C) 2019 Simon D.Levy

   MIT License
 */

public class Takeoff {

    // Target params
    static final double ALTITUDE_TARGET = 10;

    // PID params
    static final double ALT_P = 1.0;
    static final double VEL_P = 1.0;
    static final double VEL_I = 0;

    public static void main(String [] args)
    {
        // initial conditions
        double tprev = 0;
        double u = 0;

        // Create PID controller
        AltitudePidController pid  = new AltitudePidController(ALTITUDE_TARGET, ALT_P, VEL_P, VEL_I);

        // Create a multicopter simulation
        Multicopter copter = new Multicopter();

        // Start the simulation
        copter.start();

        // Loop until user hits the stop button
        while (true) {

            // Get time from sim
            double t = copter.getTime();

            // Negative time means quit
            if (t < 0) break;

            // Get state from sim
            double [] state = copter.getVehicleState();

            // Extract altitude and its first derivative from state.  Altitude
            // is in NED coordinates, so we negate it to use as input to PID
            // controller.
            double z = -state[Multicopter.STATE_Z];
            double dz = -state[Multicopter.STATE_DZ];

            // Compute vertical climb rate as first difference of altitude over time
            if (t > tprev) {

                // Use temporal first difference to compute vertical velocity
                double dt = t - tprev;

                // Get correction from PID controller
                u = pid.u(z, dz, dt);

                // Constrain correction to [0,1] to represent motor value
                u = Math.max(0., Math.min(1., u));
            }

            // Set motor values in sim
            copter.setMotors(ones(u, 4));

            // Update
            tprev = t;

            // Yield to Multicopter thread
            try {
                Thread.sleep(1);
            }
            catch (Exception e) {
            }

        }
    }

    // Helper function like Matlab/Numpy ones()
    static double [] ones(double x, int n)
    {
        double []  v = new double [n];
        for (int i=0; i<n; ++i) {
            v[i] = x;
        }
        return v;
    }
}